python/variable_init.cpp

```diff
@@ -0,0 +1,338 @@
+      std::terminate();
+    }
+
+    auto values_array =
+        Values(make_element_array<T>(dims, values, actual_unit));
+    const bool use_variances =
+        (with_variances.has_value() && *with_variances) || !variances.is_none();
+    auto variable = use_variances
+                        ? makeVariable<T>(dims, std::move(values_array),
+                                          Variances(make_element_array<T>(
+                                              dims, variances, actual_unit)))
+                        : makeVariable<T>(dims, std::move(values_array));
+    variable.setUnit(actual_unit);
+    return variable;
+  }
+};
+
+Variable make_variable(const py::object &dim_labels, const py::object &shape,
+                       const py::object &values, const py::object &variances,
+                       const units::Unit unit, DType dtype,
+                       const std::optional<bool> with_variances) {
+  const auto converted_values = parse_data_sequence(dim_labels, values);
+  const auto converted_variances = parse_data_sequence(dim_labels, variances);
+  dtype = common_dtype(converted_values, converted_variances, dtype);
+  const auto dims = build_dimensions(dim_labels, shape, converted_values,
+                                     converted_variances);
+  return core::CallDType<double, float, int64_t, int32_t, bool,
+                         scipp::core::time_point, std::string, Variable,
+                         DataArray, Dataset, Eigen::Vector3d, Eigen::Matrix3d,
+                         python::PyObject>::apply<MakeVariable>(dtype, dims,
+                                                                values,
+                                                                variances, unit,
+                                                                with_variances);
+}
+
+bool is_arg_present(const py::object &arg) { return !arg.is_none(); }
+
+template <class T> bool is_arg_present(const std::optional<T> &arg) {
+  return arg.has_value();
+}
+
+template <class A, class B>
+void ensure_mutual_exclusivity(const A &a, const std::string_view a_name,
+                               const B &b, const std::string_view b_name) {
+  if (is_arg_present(a) && is_arg_present(b)) {
+    throw std::invalid_argument(format("Passed mutually exclusive arguments '",
+                                       a_name, "' and '", b_name, "'."));
+  }
+}
+} // namespace
+
+/*
+ * It is the init method's responsibility to check that the combination
+ * of arguments is valid. Functions down the line do not check again.
+ */
+void bind_init(py::class_<Variable> &cls) {
+  cls.def(py::init([](const py::object &dim_labels, const py::object &shape,
+                      const py::object &values, const py::object &variances,
+                      const std::optional<bool> with_variances,
+                      const std::optional<units::Unit> unit,
+                      const py::object &dtype) {
```

Comment:
Is there a good justification for still supporting  `shape` and `with_variances` arguments in the constructor? Can we just support C++ version of `ones`, `zeros`, and `empty` instead, and have `Variable` only accept `values` and `variances` (not `shape` and `with_variances`)?

Would this further simplify the code?

Reply:
It would definitely simplify the code.

The reason to have them is generic code. But I am not sure this matters in practice. When deserialising, you either have data or you don't. But you would not zero-init anything, right?

Reply:
Right. You would maybe use `sc.empty` followed by a direct write to the values, but not zero-init.